knit/src/Knit/Autocompletion.hs

```diff
@@ -32,209 +37,276 @@ suggestions
      , AllConstrained ComponentPrinter components
      )
   => proxy components
+  -> Cursor
   -> T.Text
-  -> [T.Text]
-suggestions _ cmd =
+  -> [(Cursor, T.Text)]
+suggestions _ cursor cmd =
   let
-    cmd' = "(" <> cmd
-    tokens = snd $ tokenize cmd'
+    (spaceBefore, tokens) = tokenize cmd
 
     tokenBalance = \case
       TokenParenthesis bs -> withBracketSide 1 (-1) bs
       _ -> 0
-    parensBalance = sum $ map (tokenBalance . _lItem) tokens
-    cmdLines = T.splitOn "\n" cmd'
+    parensBalance = sum $ map (tokenBalance . _lItem . _twsToken) tokens
```

Comment:
I'm still unhappy about this parentheses juggling. Is there at least a task to remove it? It seems it would involve adding a new terminal `eof` to Earley, but I'd like to see it done.

Reply:
I created https://github.com/ollef/Earley/issues/37

Reply:
- I don't see how `eof` is better than this approach.
- It is possible to do without `eof` from Earley, we can extend `Knit.Tokenizer.Token` with our own eof (either directly or via an internal component), but
- It will be gone after https://issues.serokell.io/issue/AD-449 anyway.

Reply:
> I don't see how `eof` is better than this approach.

It doesn't modify the input before parsing. Here you artificially add parentheses and then remove them, for this you compute the location of the last token, and for this you have to do `T.splitOn "\n" cmd'`. It's a mess if you ask me.

> It is possible to do without eof from Earley, we can extend Knit.Tokenizer.Token with our own eof (either directly or via an internal component)

I'm not sure that'd work, would you generate an infinite sequence of `TokenEof` at the end or just one? You would need an amount equal to that of unbalanced parentheses, and at this point it's barely an improvement.

Reply:
> It will be gone after https://issues.serokell.io/issue/AD-449 anyway.

Well, I think you'll need the `eof` functionality in Early to do AD-449.